File: nodejs/tests/magnetometer/as5600_test.js
'use strict';

const { I2CTransport } = require('../../packages/periph/src/transport/i2c');
const { AS5600Full }   = require('../../packages/periph/src/chips/magnetometer/as5600');

const I2C_BUS  = parseInt(process.env.I2C_BUS  || '1',  10);
const I2C_ADDR = parseInt(process.env.I2C_ADDR  || '0x36', 16);

let passed = 0;
let failed = 0;

function checkEq(label, got, expected) {
    if (got === expected) {
        console.log('PASS', label);
        passed++;
    } else {
        console.log('FAIL %s: got %d, expected %d', label, got, expected);
        failed++;
    }
}

function checkTrue(label, condition) {
    if (condition) { console.log('PASS', label); passed++; }
    else           { console.log('FAIL', label); failed++; }
}

const transport = new I2CTransport(I2C_BUS, I2C_ADDR);

// --- Magnet status poll (60 s max at 5 Hz) ---
console.log('--- magnet status (60 s max) ---');
const _deadline = Date.now() + 60000;
while (Date.now() < _deadline) {
    const s   = transport.writeRead(Buffer.from([0x0B]), 1)[0];
    const agc = transport.writeRead(Buffer.from([0x1A]), 1)[0];
    const md = !!(s & 0x08), ml = !!(s & 0x10), mh = !!(s & 0x20);
    console.log('MD=%d ML=%d MH=%d AGC=%d', +md, +ml, +mh, agc);
    if (md) break;
    Atomics.wait(new Int32Array(new SharedArrayBuffer(4)), 0, 0, 200);
}
console.log('--- end magnet status ---');

const as5600 = Object.create(AS5600Full.prototype);
as5600._transport = transport;

// --- Magnet detection ---
checkTrue('magnet_detected', as5600.isMagnetDetected());

// --- Angle readings ---
const a = as5600.angle();
checkTrue('angle in range 0-360', a >= 0.0 && a < 360.0);

const r = as5600.angleRaw();
checkTrue('angle_raw in range 0-4095', r >= 0 && r <= 4095);

const ra = as5600.rawAngle();
checkTrue('raw_angle in range 0-4095', ra >= 0 && ra <= 4095);

const rad = as5600.rawAngleDegrees();
checkTrue('raw_angle_degrees in range 0-360', rad >= 0.0 && rad < 360.0);

// --- Diagnostics ---
checkTrue('agc non-negative', as5600.agc() >= 0);
checkTrue('magnitude non-negative', as5600.magnitude() >= 0);

// --- Status ---
const sb = as5600.statusByte();
checkTrue('status_byte valid', sb >= 0 && sb <= 255);

// --- Position configuration (volatile) ---
as5600.setZeroPosition(100);
checkEq('zero_position after set', as5600.zeroPosition(), 100);

as5600.setMaxPosition(2000);
checkEq('max_position after set', as5600.maxPosition(), 2000);

as5600.setMaxAngle(2048);
checkEq('max_angle after set', as5600.maxAngle(), 2048);

// --- Configure ---
as5600.configure(0, 0, 0, 0, 0, 0, false);
checkTrue('configure accepted', as5600.isMagnetDetected());

// --- Burn count ---
const bc = as5600.burnCount();
checkTrue('burn_count in range 0-3', bc >= 0 && bc <= 3);

transport.close();

console.log('===DONE: %d passed, %d failed===', passed, failed);
process.exit(failed === 0 ? 0 : 1);

File: python/tests/magnetometer/as5600_test.py
import time
import _testconfig as cfg
from periph.transport.i2c_micropython import I2CTransport
from periph.chips.magnetometer.as5600 import AS5600Full

from machine import I2C, Pin

passed = 0
failed = 0


def check_eq(label, got, expected):
    global passed, failed
    if got == expected:
        print('PASS', label)
        passed += 1
    else:
        print('FAIL {}: got {}, expected {}'.format(label, got, expected))
        failed += 1


def check_true(label, condition):
    global passed, failed
    if condition:
        print('PASS', label)
        passed += 1
    else:
        print('FAIL', label)
        failed += 1


i2c = I2C(cfg.I2C_ID, sda=Pin(cfg.SDA), scl=Pin(cfg.SCL), freq=cfg.FREQ)
transport = I2CTransport(i2c, cfg.ADDR)
as5600 = AS5600Full(transport)

# --- Magnet detection ---
check_true('magnet_detected', as5600.is_magnet_detected())

# --- Angle readings ---
a = as5600.angle()
check_true('angle in range 0-360', a >= 0.0 and a < 360.0)

r = as5600.angle_raw()
check_true('angle_raw in range 0-4095', r >= 0 and r <= 4095)

ra = as5600.raw_angle()
check_true('raw_angle in range 0-4095', ra >= 0 and ra <= 4095)

rad = as5600.raw_angle_degrees()
check_true('raw_angle_degrees in range 0-360', rad >= 0.0 and rad < 360.0)

# --- Diagnostics ---
check_true('agc non-negative', as5600.agc() >= 0)
check_true('magnitude non-negative', as5600.magnitude() >= 0)

# --- Status ---
sb = as5600.status_byte()
check_true('status_byte valid', sb >= 0 and sb <= 255)

# --- Position configuration (volatile) ---
as5600.set_zero_position(100)
check_eq('zero_position after set', as5600.zero_position(), 100)

as5600.set_max_position(2000)
check_eq('max_position after set', as5600.max_position(), 2000)

as5600.set_max_angle(2048)
check_eq('max_angle after set', as5600.max_angle(), 2048)

# --- Configure ---
as5600.configure(pm=0, hyst=0, outs=0, pwmf=0, sf=0, fth=0, wd=False)
check_true('configure accepted', as5600.is_magnet_detected())

# --- Burn count ---
bc = as5600.burn_count()
check_true('burn_count in range 0-3', bc >= 0 and bc <= 3)

print('===DONE: {} passed, {} failed==='.format(passed, failed))
